scripts/vault-push-approle-envs.nix

```diff
@@ -0,0 +1,109 @@
+# Generate and push approles to vault
+{ writeShellScriptBin, jq, vault, coreutils, lib }:
+# Inputs: a flake with `nixosConfigurations`
+
+# Usage:
+# apps.x86_64-linux.vault-push-approle-envs = { type = "app"; program = "${pkgs.vault-push-approle-envs self}/bin/vault-push-approle-envs"; }
+
+{ nixosConfigurations ? { }, ... }: rec {
+  overrideable = final: {
+    hostNameOverrides = { };
+    getHostName = { attrName, config, ... }:
+      final.hostNameOverrides.${attrName} or "${config.networking.hostName}.${config.networking.domain}";
+  };
+
+  __toString = self:
+    let
+
+      final = lib.fix self.overrideable;
+
+      # The script that writes the approle to the vault server
+      pushApproleEnv =
+        { approleName, vaultAddress, environmentFile, ... }@params:
+        let
+          hostname = final.getHostName params;
+
+          push = ''
+            ${./vault-get-approle-env.sh} ${approleName} | ssh "${hostname}" ''${SSH_OPTS:-} "sudo mkdir -p ${
+              builtins.dirOf environmentFile
+            }; sudo -n tee ${environmentFile} >/dev/null"
+          '';
+        in ''
+          export VAULT_ADDR="${vaultAddress}"
+
+          if [[ $# -eq 0 ]] || [[ " $@ " =~ " ${approleName} " ]]; then
+            # If we don't get any arguments, or the current approle name is in the arguments list, push it
+            echo "Uploading ${approleName} to ${hostname}"
+            set -x
+            ${push}
+            set +x
+          fi
+        '';
+
+      # Get all approles for vault-secrets in configuration
+      approleParamsForMachine = attrName: cfg:
+        let
+          vs = cfg.config.vault-secrets;
+          prefix = lib.optionalString (!isNull vs.approlePrefix)
+            "${vs.approlePrefix}-";
+        in builtins.attrValues (builtins.mapAttrs (name: secret:
+          builtins.removeAttrs (vs // secret // {
+            approleName = "${prefix}${name}";
+            inherit name attrName;
+            inherit (cfg) config;
+          }) [ "__toString" "secrets" ]) vs.secrets);
```

Comment:
Some of the definitions here are the same as in the other script, it would be good to export them to a common nix file. Then it would be more readable, because we wouldn't have to check if there are subtle differences between the definitions, and it would be easier to maintain, because we wouldn't have to update the definitions in two places.

Reply:
Good idea, will do